fawltydeps/utils.py

```diff
@@ -23,26 +38,154 @@ def version() -> str:
+    3. Allow the traversal to adjust itself while in progress. For example,
+       while traversing /some/dir we want to be able "ignore" a directory
+       (including /some/dir or any parent) from (further) traversal. Likewise,
+       we should be able to _add_ more directories to the traversal. (For a
+       given traversal, re-adding a directory that was already traversed shall
+       have no effect, if you want to re-traverse then setup a _new_ traversal.)
+
+    However, we _do_ assume that the directory structures being traversed remain
+    unchanged during traversal. I.e. a directory
+    """
+
+    to_traverse: Set[Path] = field(default_factory=set)
+    to_ignore: Set[DirId] = field(default_factory=set)  # includes already-traversed
+    attached: Dict[DirId, List[T]] = field(default_factory=dict)
+
+    def add(self, dir_path: Path, attach_data: Optional[T] = None) -> None:
+        """Add one directory to this traversal, optionally w/attached data.
+
+        - Any attached data will be supplied to the given directory and its
+          subdirectories when it is being .traverse()d.
+        - A directory can be added multiple times with different attached data;
+          all the attached data will be supplied (in the order it was added) by
+          .traverse().
+        - A parent directory and a child directory may both be added with
+          different data attached. The child directory will be supplied both the
+          parent's data (first) and its own data (last) by .traverse().
+        - No matter how many times a directory is added, it will still only be
+          traversed _once_. If a directory has already been traversed by this
+          instance, it will _not_ be re-traversed.
+        """
+        if not dir_path.is_dir():
+            raise NotADirectoryError(dir_path)
+        dir_id = DirId.from_path(dir_path)
+        self.to_traverse.add(dir_path)
+        if attach_data:
+            self.attached.setdefault(dir_id, []).append(attach_data)
+
+    def ignore(self, dir_path: Path) -> None:
+        """Ignore a directory in future traversal.
+
+        The given directory or its subdirectories will _not_ be traversed
+        (although explicitly .add()ed subdirectories _will_ be traversed).
+        """
+        self.to_ignore.add(DirId.from_path(dir_path))
+
+    def traverse(self) -> Iterator[Tuple[Path, Set[Path], Set[Path], List[T]]]:
+        """Perform the traversal of the added directories.
+
+        For each directory traverse, yield a 4-element tuple consisting of:
+        - The path to the current directory being traversed.
+        - An unordered iterator of all (immediate) subdirectories in the current
+          directory. This is NOT a mutable list, as you might expect from
+          os.walk(); instead, to prevent traversing into a subdirectory, you can
+          call .ignore() with the relevant subdirectory.
+        - An unordered iterator of all files in the current directory
```

Comment:
Oops, I forgot to update the docstring after a late refactoring. These _could_ have been lazy iterators instead of sets, but then the caller would have to take extra care to make sure they were iterated at most _once_. I'd rather keep things simpler although arguably too much is computed in the case where the caller ends up not referencing these sets.

Will fix the docstring.

Reply:
Fixed docstring in c3fd4d6.